notifyme-sdk-backend/notifyme-sdk-backend-ws/src/main/java/ch/ubique/notifyme/sdk/backend/ws/util/CryptoUtilV3.java

```diff
@@ -0,0 +1,320 @@
+        byte[] msg_dec = decryptInternal(ibeCiphertext, secretKeyForIdentity, identity);
+        if (msg_dec == null) {
+            throw new RuntimeException("Health Authority could not verify Trace");
+        }
+    }
+
+    public List<TraceKey>
+    createTraceV3ForUserUpload(UserUploadPayloadOuterClass.UserUploadPayload userUpload) {
+        var traceKeys = new ArrayList<TraceKey>();
+        for (UserUploadPayloadOuterClass.UploadVenueInfo venueInfo : userUpload.getVenueInfosList()) {
+            if (!venueInfo.getFake()) {
+                byte[] identity =
+                        cryptoHashSHA256(
+                                concatenate(
+                                        "CN-ID".getBytes(StandardCharsets.US_ASCII),
+                                        venueInfo.getPreId().toByteArray(),
+                                        intToBytes(3600),
+                                        longToBytes(venueInfo.getIntervalStartMs() / 1000),
+                                        venueInfo.getTimeKey().toByteArray()));
+
+                G1 secretKeyForIdentity = keyDer(this.mskFr, identity);
+
+                byte[] nonce = createNonce();
+                NotifyMeAssociatedDataOuterClass.NotifyMeAssociatedData countryData =
+                        NotifyMeAssociatedDataOuterClass.NotifyMeAssociatedData.newBuilder()
+                                .setCriticality(NotifyMeAssociatedDataOuterClass.EventCriticality.LOW)
+                                .setVersion(1)
+                                .build();
+                byte[] encryptedAssociatedData =
+                        this.encryptAssociatedData(
+                                venueInfo.getNotificationKey().toByteArray(),
+                                "",
+                                countryData.toByteArray(),
+                                nonce,
+                                venueInfo.getIntervalStartMs(),
+                                venueInfo.getIntervalEndMs());
+
+                var traceKeyV3 = new ch.ubique.notifyme.sdk.backend.model.tracekey.v3.TraceKey();
+                traceKeyV3.setVersion(3);
+                traceKeyV3.setCipherTextNonce(nonce);
+                traceKeyV3.setEncryptedAssociatedData(encryptedAssociatedData);
+                traceKeyV3.setDay(
+                        Instant.ofEpochMilli(venueInfo.getIntervalStartMs()).truncatedTo(ChronoUnit.DAYS));
+                traceKeyV3.setIdentity(identity);
+                traceKeyV3.setSecretKeyForIdentity(secretKeyForIdentity.serialize());
+                traceKeys.add(traceKeyV3);
+
+                // verifyTrace
+                int NONCE_LENGTH = 32;
+                byte[] msg_orig = createNonce(NONCE_LENGTH);
+                IBECiphertext ibeCiphertext = encryptInternal(this.mpkG2, identity, msg_orig);
+                byte[] msg_dec = decryptInternal(ibeCiphertext, secretKeyForIdentity, identity);
+                if (msg_dec == null) {
```

Comment:
Why do we need to check this on every insert? And if we do, should we not also check that msg_dec matches msg_orig?

Reply:
Resolved: See [https://github.com/SwissCovid/swisscovid-cn-backend/pull/2/commits/d7c61ba3c47b16a9653d3209b1186c3de974c66c](https://github.com/SwissCovid/swisscovid-cn-backend/pull/2/commits/d7c61ba3c47b16a9653d3209b1186c3de974c66c)
We need this verification step according to section 4.5 "Initiating Presence Notification" step 2.4 of the CrowdNotifier Technical Specification Documentation.